module/zfs/dsl_dir.c

```diff
@@ -2224,6 +2242,86 @@ dsl_dir_is_zapified(dsl_dir_t *dd)
 	return (doi.doi_type == DMU_OTN_ZAP_METADATA);
 }
 
+void
+dsl_dir_livelist_open(dsl_dir_t *dd, uint64_t obj)
+{
+	objset_t *mos = dd->dd_pool->dp_meta_objset;
+	ASSERT(spa_feature_is_active(dd->dd_pool->dp_spa,
+	    SPA_FEATURE_LIVELIST));
+	dsl_deadlist_open(&dd->dd_livelist, mos, obj);
+	bplist_create(&dd->dd_pending_allocs);
+	bplist_create(&dd->dd_pending_frees);
+}
+
+void
+dsl_dir_livelist_close(dsl_dir_t *dd)
+{
+	dsl_deadlist_close(&dd->dd_livelist);
+	bplist_destroy(&dd->dd_pending_allocs);
+	bplist_destroy(&dd->dd_pending_frees);
+}
+
+void
+dsl_dir_remove_livelist(dsl_dir_t *dd, dmu_tx_t *tx, boolean_t total)
+{
+	uint64_t obj;
+	dsl_pool_t *dp = dmu_tx_pool(tx);
+	spa_t *spa = dp->dp_spa;
+	if (!dsl_deadlist_is_open(&dd->dd_livelist))
+		return;
+	livelist_condense_entry_t to_condense = spa->spa_to_condense;
+
+	/*
+	 * If the livelist being removed is set to be condensed, stop the
+	 * condense zthr and indicate the cancellation in the spa_to_condense
+	 * struct in case the condense no-wait synctask has already started
+	 */
+	zthr_t *ll_condense_thread = spa->spa_livelist_condense_zthr;
+	if (ll_condense_thread != NULL &&
+	    (to_condense.ds != NULL) && (to_condense.ds->ds_dir == dd)) {
+			/*
+			 * We use zthr_wait_cycle_done instead of zthr_cancel
+			 * because we don't want to destroy the zthr, just have
+			 * it skip it's current task.
+			 */
+			spa->spa_to_condense.cancelled = B_TRUE;
+			zthr_wait_cycle_done(ll_condense_thread);
+			/*
+			 * If we've returned from zthr_wait_cycle_done without
+			 * clearing the to_condense data structure it's either
+			 * because the no-wait synctask has started (which is
+			 * indicated by 'syncing' field of to_condense) and we
+			 * can expect it to clear to_condense on its own.
+			 * Otherwise, we returned before the zthr ran. The
+			 * checkfunc will now fail as cancelled == B_TRUE so we
+			 * can safely NULL out ds, allowing a different dir's
+			 * livelist to be condensed.
+			 *
+			 * We can be sure that the to_condense struct will not
+			 * be repopulated at this stage because both this
+			 * function and dsl_livelist_try_condense execute in
+			 * syncing context.
+			 */
+			if ((spa->spa_to_condense.ds != NULL) &&
+			    !spa->spa_to_condense.syncing) {
+				dmu_buf_rele(spa->spa_to_condense.ds->ds_dbuf,
+				    spa);
+				spa->spa_to_condense.ds = NULL;
+			}
+	}
+
+	dsl_dir_livelist_close(dd);
+	int err = zap_lookup(dp->dp_meta_objset, dd->dd_object,
+	    DD_FIELD_LIVELIST, sizeof (uint64_t), 1, &obj);
+	ASSERT3U(err, !=, ENOENT);
```

Comment:
For the `ECKSUM` handling - would you recommend just making sure the callers are protected (bail out it err != 0) or do you think I should make changes to propagate the `ECKSUM` error further up the stack?

Reply:
The way we've handled this elsewhere is to return an error up the stack whenever we can.  This is probably the most important for the import path and should be pretty straight forward.  The ZTS extreme rewind are expected to encounter this periodically.  For the other places where we can't bail, if there's a failure that can't be handled without causing pool damage then there's a VERIFY.  Which isn't ideal, but addressing that is a project in and of itself.

Reply:
Cool - thanks for the suggestions. I'll add the error return in `dsl_dir_hold_obj` (which is what you mean by the import path, right?)

Reply:
For the import path I was worried about `spa_ld_get_props()`, but everything looks good there now!  For `dsl_dir_hold_obj()` actually, I think it may be fine.  What would be the consequence of not calling `dsl_dir_livelist_open()` here?

Reply:
It would just not use the livelist feature, which isn't a big deal. Though if it were to later successfully open the livelist, there would be issues will some ALLOCs not being tracked.

Reply:
Brian and I talked offline about this and agreed that returning the err in this function is sufficient to prevent this livelist non-determinism possibility.